Mathlib/Tactic/Recall.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2023 Mac Malone. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mac Malone, Kyle Miller
+-/
+import Lean.Elab.MutualDef
+import Std.Tactic.OpenPrivate
+
+/-!
+# `recall` command
+-/
+
+namespace Mathlib.Tactic.Recall
+
+/--
+The `recall` command redeclares a previous definition for illustrative purposes.
+This can be useful for files that give an expository account of some theory in Lean.
+
+The syntax of the command mirrors `def`, so all the usual bells and whistles work.
+```
+recall HasFDerivAtFilter (f : E → F) (f' : E →L[𝕜] F) (x : E) (L : Filter E) :=
+  (fun x' => f x' - f x - f' (x' - x)) =o[L] fun x' => x' - x
```

Comment:
Could this use an example that is likely to be available from no imports (or only the ones used by this file)? Alternatively, this example could include an import of `Mathlib.Tactic.Recall` itself along with whatever file you need to get this fderiv example.

Reply:
@digama0 Is the goal just to have a simpler example? I choose that one because it was complex, to demonstrate that everything works as expected (and also because it was the motivating example in the thread). However, if the goal is just to have something people can copy and paste, maybe this would work:
```lean
recall Function.comp {α} {β} {δ} (f : β → δ) (g : α → β) : α → δ :=
  fun x => f (g x)
```
What do you think?

Reply:
You could still demonstrate things like notations, variables and autoimplicits using something from lean or std. My reading was that this was what you wanted to show off with this example.

Reply:
@digama0 Do you have any suggestions on a good example? Finding sophisticated notation in both types and definition of a non-proposition is a lot harder with Lean/Std as most definitions there have relatively simple types. And the ones that do not are generally relatively obscure definitions that do not make for great examples.

Reply:
how about:
```lean
recall List.cons_append (a : α) (as bs : List α) : (a :: as) ++ bs = a :: (as ++ bs) := rfl
```

Reply:
@digama0 That does have a lot of nice notation and some auto-implicits, but it is still a theorem. Thus, it does not really demonstrate the utility of being able to restate value of a definition.  I guess the question is which is worth more in this first example?

Reply:
well, the fact that the theorem is proved by rfl is demonstrated by the `recall`, even though "value typechecks" is pulling more weight than "value is defeq to original definition" here.

Reply:
Okay, if you think that is good enough, I will use it.